specification/dartLangSpec.tex

```diff
@@ -10474,84 +10508,158 @@ \subsection{Assignment}
+
+Otherwise
+(\commentary{when \id{} is not a local variable}),
+a lexical lookup of \code{\id=}
+(\ref{lexicalLookup})
+is performed from the location of \id.
+
+\begin{itemize}
+\item
+  When the lexical lookup yields a declaration $D$,
+  it is guaranteed to be a setter
+  (\commentary{that may be explicit or induced implicitly by a variable})
+  because other declarations do not have a name
+  of the form \code{\id=}.
+
+  If $D$ is the declaration of a static setter in class $C$
+  then $a$ is treated as
+  (\ref{notation})
+  the assignment \code{$C$.\id{} = $e$}.
+
+  \commentary{%
+    Further analysis as well as evaluation of \code{$C$.\id{} = $e$}
+    proceeds as specified elsewhere.%
+  }
+
+  Otherwise, a compile-time error occurs,
+  unless the static type of $e$ is assignable to the parameter type of $D$.
+\item
+  When the lexical lookup yields a member signature,
+  $a$ is treated as
+  (\ref{notation})
+  \code{\THIS.\id{} = $e$}.
+
+  \commentary{%
+    This occurs when no declaration named \code{\id=} is in scope,
+    but $a$ occurs in the body of an instance member declaration,
+    and the interface of the enclosing class has a setter named \code{\id=}.
+    Both the static analysis and evaluation proceeds with
+    \code{\THIS.\id{} = $e$},
+    so there is no need to further specify the treatment of $a$.%
+  }
+\item
+  \commentary{%
+    The lexical lookup can never yield an import prefix,
+    because they never have a name of the form \code{\id=}.%
+  }
+\end{itemize}
 
 \LMHash{}%
-It is a compile-time error if an assignment of the form \code{$v$ = $e$} occurs
-inside a top level or static function (be it function, method, getter, or setter) or variable initializer,
-and there is neither a mutable local variable declaration with name $v$
-nor a setter declaration with name \code{$v$=} in the lexical scope enclosing the assignment.
+In all cases
+(\commentary{whether or not \id{} is a local variable, etc.}),
+the static type of $a$ is the static type of $e$.
 
 \LMHash{}%
-Evaluation of an assignment $a$ of the form \code{$v$ = $e$}
+Evaluation of an assignment $a$ of the form \code{\id{} = $e$}
 proceeds as follows:
-%% TODO(eernst): $d$ is defined ambiguously: both getter & setter may exist.
-Let $d$ be the innermost declaration whose name is $v$ or \code{$v$=}, if it exists.
-It is a compile-time error if $d$ denotes
-a prefix object, type declaration, or function declaration.
 
-\LMHash{}%
-If $d$ is the declaration of a local variable, the expression $e$ is evaluated to an object $o$.
-Then, the variable $v$ is bound to $o$.
-If no error occurs, the value of the assignment expression is $o$.
+\begin{itemize}
+\item
+  In the case where \id{} is a local variable $v$
+  (\commentary{which may be a formal parameter}),
+  the expression $e$ is evaluated to an object $o$,
+  and the variable $v$ is bound to $o$.
+  If no error occurs, the value of $a$ is $o$.
```

Comment:
Just say: 

> Then $a$ evaluates to the object $o$ (\ref{expressionEvaluation}).

Errors are stated elsewhere to automatically interrupt evaluation, only mention them when they are introduced or when you need to handle them, otherwise they propagate automatically.

Reply:
Done.